src/AAA_Loader.py

```diff
@@ -0,0 +1,187 @@
+import inspect
+import logging
+import os
+import sys
+from collections import defaultdict, namedtuple
+from importlib import import_module
+from importlib.abc import Loader, MetaPathFinder
+from importlib.machinery import ModuleSpec
+from importlib.util import spec_from_loader
+from types import ModuleType
+from typing import Dict, Iterable, List, Optional
+
+__version__ = "0.1"
+HELLO_MY_NAME_IS = "Loader"
+MODS_FLAG = "loadmods"
+
+Mod = namedtuple("Mod", "name path module_path")
+
+log = logging.getLogger("AAA_Loader")
+
+
+# Readonly list type for stopping duplicate mod list entries in crashlogs
+class ReadOnlyList(list):  # type: ignore[type-arg]
+    def append(*args, **kwargs):  # type: ignore[no-untyped-def]
+        pass
+
+
+class AssetLoader:
+    def __init__(self, path: str) -> None:
+        self.base_path = path
+
+    def get_asset(self, path: List[str]) -> List[str]:
+        assert path
+        assert isinstance(path, list)
+        assert isinstance(path[0], str)
+
+        desired_path = os.path.join(self.base_path, *path)
+        relative_path = os.path.relpath(desired_path, os.path.abspath("rl_data"))
+
+        return relative_path.split(os.sep)
+
+
+class ConstantImporter(MetaPathFinder, Loader):
+    def __init__(self, fullname: str, module: ModuleType):
+        self.fullname = fullname
+        self.module = module
+
+    def find_spec(cls, fullname: str, path=None, target=None) -> Optional[ModuleSpec]:  # type: ignore[no-untyped-def]
+        if fullname == cls.fullname:
+            return spec_from_loader(fullname, cls, origin=cls.fullname)
+        return None
+
+    def create_module(cls, spec: ModuleSpec) -> ModuleType:
+        return cls.module
+
+    def exec_module(cls, module=None):  # type: ignore[no-untyped-def]
+        pass
+
+    def get_code(cls, fullname):  # type: ignore[no-untyped-def]
+        return None
+
+    def get_source(cls, fullname):  # type: ignore[no-untyped-def]
+        return None
+
+    def is_package(cls, fullname):  # type: ignore[no-untyped-def]
+        return False
+
+
+RiftWizard: Optional[ModuleType] = None
+
+
+def patch_imports(stack: List["inspect.FrameInfo"]) -> None:
+    global RiftWizard
+    # Try to find RiftWizard module á la
+    # https://github.com/RiftWizard-Modders/RiftWizard-Modding/wiki/Mod-Setup#importing-riftwizardpy
+    frm = stack[-1]
```

Comment:
this may need adjusting if it's no longer ran directly from the module, since stack[-1] would point to the caller of patch_imports. clay linked me the following snippet which should be more robust:

```
####################################################
# Importing RiftWizard.py                          |
# Credit to trung on discord                       |
#                                                  |
#----------------------------------------------    |
import inspect #                                   |
def get_RiftWizard(): #                            |
    # Returns the RiftWizard.py module object      |
    for f in inspect.stack()[::-1]: #              |
        if "file 'RiftWizard.py'" in str(f): #     |
            return inspect.getmodule(f[0]) #       |
     #                                             |
    return inspect.getmodule(f[0]) #               |
#                                                  |
RiftWizard = get_RiftWizard() #                    |
#                                                  |
#                                                  |
####################################################
```